### PR TITLE
Fix wrong pad num in rich progress bar

### LINKED ISSUE #12113: Wrong pad number of space in rich progerss bar get train description func
## 🐛 Bug
https://github.com/PyTorchLightning/pytorch-lightning/blob/6fcfd8ce91c5e6d5721243c1d8c7606ee3079460/pytorch_lightning/callbacks/progress/rich_progress.py#L425-L434
There is some mistake in calcultion of `required_padding`, it should be 
```python
required_padding = len(self.validation_description) - len(train_description)
```
Currently, we can also see flickering when `current_epoch` > 9
### Additional context

By the way, how about display the max epoch in train description instead of pad space?
The code will change to
```python
    def _get_train_description(self, current_epoch: int) -> str:
        train_description = f"Epoch {current_epoch + 1}"
        if self.trainer.max_epochs is not None and self.trainer.max_epochs < 1000:
            train_description += f"/{self.trainer.max_epochs}"
        if len(self.validation_description) > len(train_description):
            # Padding is required to avoid flickering due of uneven lengths of "Epoch X"
            # and "Validation" Bar description
            required_padding = len(self.validation_description) - len(train_description)
            train_description += " " * required_padding
        return train_description
```
And, the progress bar will look like
截屏2022-02-25 18 31 49

cc @SeanNaren @kaushikb11

### PR DESCRIPTION
**Pull request by shenmishajing**
## What does this PR do?

Fixes #12113

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 1]
```diff
@@ -423,14 +423,14 @@ def on_predict_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, da
...
+            train_description += f"/{self.trainer.max_epochs}"
         if len(self.validation_description) > len(train_description):
             # Padding is required to avoid flickering due of uneven lengths of "Epoch X"
             # and "Validation" Bar description
-            num_digits = len(str(current_epoch))
-            required_padding = (len(self.validation_description) - len(train_description) + 1) - num_digits
-            for _ in range(required_padding):
-                train_description += " "
+            required_padding = len(self.validation_description) - len(train_description)
+            train_description += " " * required_padding
```
Comment (carmocca) [inline pytorch_lightning/callbacks/progress/rich_progress.py:433]: Let's use fstring padding directly

https://saralgyaan.com/posts/f-string-in-python-usage-guide/#Python_f-strings_padding

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 2]
```diff
@@ -423,14 +423,14 @@ def on_predict_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, da
         self.refresh()
 
     def _get_train_description(self, current_epoch: int) -> str:
-        train_description = f"Epoch {current_epoch}"
+        train_description = f"Epoch {current_epoch + 1}"
```
[the remark this answers is not in the document]
Comment (shenmishajing) [reply pytorch_lightning/callbacks/progress/rich_progress.py:426]: But, it will start from 0/24, and end with 23/24, if we do not change this. I think it's a little bit weird.

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 1 above)
Comment (shenmishajing) [reply pytorch_lightning/callbacks/progress/rich_progress.py]: fix this.

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/callbacks/progress/rich_progress.py:426]: In that case, the right hand side should have `-1`

**Comment by carmocca** [comment]
Can you also update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md) and add a test for this?

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/callbacks/progress/rich_progress.py [hunk 3]
```diff
@@ -451,13 +451,12 @@ def on_predict_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, da
 
     def _get_train_description(self, current_epoch: int) -> str:
         train_description = f"Epoch {current_epoch}"
+        if self.trainer.max_epochs is not None and self.trainer.max_epochs < 1000:
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/progress/rich_progress.py:454]: Why the `< 1000` filter?

File: tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py [hunk 4]
```diff
@@ -401,3 +401,16 @@ def test_step(self, batch, batch_idx):
...
+
+
+@RunIf(rich=True)
+def test_rich_progress_bar_padding(tmpdir):
+
+    progress_bar = RichProgressBar()
+    model = BoringModel()
+
+    trainer = Trainer(default_root_dir=tmpdir, callbacks=progress_bar, fast_dev_run=True)
+    trainer.fit(model)
```
Comment (carmocca) [inline tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py:413]: It's not required that we `fit` a model for this test. It doesn't even need to have `RunIf(rich=True)` if instead we unit-test the `_get_train_description` method

**Review by carmocca — CHANGES_REQUESTED** [review]

File: tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py (hunk 4 above)
Comment (kaushikb11) [reply tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py:413]: But it would require class attributes like progress, etc to figure out the description for the progress bar.

Comment (carmocca) [reply tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py:413]: That's fine. See for example:

```py
def test_rich_progress_bar_padding():
    progress_bar = RichProgressBar()
    trainer = Mock()
    trainer.max_epochs = 1
    progress_bar._trainer = trainer

    train_description = progress_bar._get_train_description(current_epoch=0)
    assert "Epoch 0/0" in train_description
    assert len(progress_bar.validation_description) == len(train_description)
```

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/callbacks/progress/rich_progress.py [hunk 5]
```diff
@@ -451,13 +451,12 @@ def on_predict_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, da
...
+        if self.trainer.max_epochs is not None:
+            train_description += f"/{self.trainer.max_epochs - 1}"
         if len(self.validation_description) > len(train_description):
             # Padding is required to avoid flickering due of uneven lengths of "Epoch X"
             # and "Validation" Bar description
-            num_digits = len(str(current_epoch))
-            required_padding = (len(self.validation_description) - len(train_description) + 1) - num_digits
-            for _ in range(required_padding):
-                train_description += " "
+            train_description = f"{train_description:{len(self.validation_description)}}"
```
Comment (Borda) [inline src/pytorch_lightning/callbacks/progress/rich_progress.py:459]: does this need to be fstring when there is just variable printing?

File: tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py (hunk 4 above)
Comment (kaushikb11) [reply tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py]: Done!

File: src/pytorch_lightning/callbacks/progress/rich_progress.py (hunk 5 above)
Comment (kaushikb11) [reply src/pytorch_lightning/callbacks/progress/rich_progress.py:459]: @Borda Wdym?
It will be an argument to the `TextColumn` https://rich.readthedocs.io/en/stable/progress.html#columns

Comment (Borda) [reply src/pytorch_lightning/callbacks/progress/rich_progress.py:459]:
```suggestion
            train_description = repr(train_description:{len(self.validation_description)})
```

File: tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py [hunk 6]
```diff
@@ -403,11 +403,8 @@ def test_step(self, batch, batch_idx):
     assert pbar.calls["test"] == []
 
 
-def test_rich_progress_bar_padding(monkeypatch):
-    import pytorch_lightning.callbacks.progress.rich_progress as imports
-
-    monkeypatch.setattr(imports, "_RICH_AVAILABLE", True)
```
Comment (carmocca) [inline tests/tests_pytorch/callbacks/progress/test_rich_progress_bar.py]: Why remove the monkeypatch?

File: src/pytorch_lightning/callbacks/progress/rich_progress.py (hunk 5 above)
Comment (carmocca) [reply src/pytorch_lightning/callbacks/progress/rich_progress.py:459]: It's not just variable printing, it's using the padding argument of f-strings as indicated in the comment above.

Using `repr` here is not equivalent.

**Comment by kaushikb11** [comment]
@Borda Seems like I can't resolve the conflicts as `https://github.com/shenmishajing/pytorch-lightning.git` doesn't exist.

Do you know a solution for this?

**Comment by carmocca** [comment]
@kaushikb11 I suggest copying the [patch](https://patch-diff.githubusercontent.com/raw/Lightning-AI/lightning/pull/12125.patch) into a branch of your own and opening a new PR that closes this one
